wasm_node_api/wasm_node_api.cc

```diff
@@ -0,0 +1,886 @@
+    }
+  #if __WORDSIZE == 64
+  } else if constexpr(std::is_pointer<kind<Result>>::value) { // Native pointer
+    return (kind<Result>) v.i64();
+  #endif
+  } else {
+    return napiCast<kind<Result>>(v);
+  }
+};
+
+template<typename Result>
+constexpr w::Val napiWrap(kind<Result> v) {
+  if constexpr(std::is_same<kind<Result>, float>::value) {
+    return w::Val((float) v);
+  } else if constexpr (std::is_same<kind<Result>, double>::value) {
+    return w::Val((double) v);
+  #if __WORDSIZE == 64
+  } else if constexpr(std::is_pointer<kind<Result>>::value) { // Native pointer
+      return w::Val(static_cast<int64_t>((size_t) v));
+  #endif
+  } else {
+    return w::Val(static_cast<int32_t>((size_t) v));
+  }
+}
+
+template<typename Result, typename... Args>
+struct napiApply {
+  template<size_t... Indices>
+  static constexpr auto apply(
+    kind<Result> (*fn)(kind<Args>...),
+    const w::Context* ctx,
+    const w::Val args[],
+    std::index_sequence<Indices...>
+  ) -> kind<Result> {
+    if constexpr(std::is_void<kind<Result>>::value) {
+      fn(napiUnwrap<Args>(ctx, args[Indices])...);
+    } else {
+      return fn(napiUnwrap<Args>(ctx, args[Indices])...);
+    }
+  }
+};
+
+template<typename Result, typename... Args>
+struct napiCall {
+private:
+  using fp = kind<Result> (*)(kind<Args>...);
+public:
+  template<fp Fn>
+  static void const call(
+    const w::Context* ctx, const w::Val args[], w::Val results[]
+  ) {
+    if constexpr(std::is_void<kind<Result>>::value) {
+      napiApply<Result, Args...>::apply(Fn, ctx, args,
+                                  std::index_sequence_for<Args...>{});
+    } else {
+      results[0] = napiWrap<Result>(
+        napiApply<Result, Args...>::apply(Fn, ctx, args,
+                                          std::index_sequence_for<Args...>{}));
+    }
+  }
+};
+
+template<typename Result, typename... Args>
+struct napiBind {
+  template<kind<Result> (*const Fn)(kind<Args>...)>
+  static void bind(v8::Isolate* isolate, const char *name) {
+    w::Func fn = exportNative<Result, Args...>(
+        (callback) napiCall<Result, Args...>::template call<Fn>);
+    w::RegisterEmbedderBuiltin(isolate, "napi_unstable", name, fn);
+  }
+};
+
+// Functions to deal with copying strings
+size_t get_string_length(const char* cString) {
+  return strlen(cString);
+}
+
+void copy_string_to_wasm(const char* cString, char* wasmString) {
+  strcpy(wasmString, cString);
+}
+
+// Functions for dealing with napi modules
+napi_module* create_napi_module() {
+  return (napi_module*) malloc(sizeof(napi_module));
+}
+
+void set_napi_module_nm_version(napi_module* mod, int nm_version) {
+  mod->nm_version = nm_version;
+}
+
+void set_napi_module_nm_flags(napi_module* mod, unsigned int nm_flags) {
+  mod->nm_flags = nm_flags;
+}
+
+void set_napi_module_nm_filename(napi_module* mod, const char* nm_filename) {
+  mod->nm_filename = nm_filename;
+}
+
+// TODO(ohadrau): Is this a safe operation?
```

Comment:
What would make it not safe? Is it that you are passing in a function pointer?

Reply:
The concern here is that the function pointer that gets passed in is not a valid function pointer, since it refers to an index in the WASM indirect function table. Since we theoretically control 100% of the use of this function pointer, we're able to get away with using this hack but it is a big assumption to make in my opinion.